src/app/modules/spcp/sp.oidc.client.cache.ts

```diff
@@ -0,0 +1,211 @@
+
+  /**
+   * Method to retrieve base client from cache
+   * Triggers a refresh
+   * @async
+   * @returns Base client
+   * @throws error if refresh does not complete within 10s
+   */
+  async getBaseClient(): Promise<BaseClient> {
+    const baseClient = this._cache.get<BaseClient>(BASE_CLIENT_NAME)
+    if (!baseClient) {
+      const { baseClient } = await timeout(this.refresh(), 10000)
+      return baseClient
+    }
+    return baseClient
+  }
+
+  /**
+   * Method to create a promise to fetch NDI's public keys and
+   * discover the well known endpoint to construct the base client,
+   * and store NDI's public key and base client in cache
+   * Sets `expiry` key in cache with TTL of 1hour to attempt refresh ahead
+   * Will retry infinite number of times until success, with each retry consisting of
+   * at most 3 back-to-back attempts (retryPromiseThreeAttempts), and
+   * 10s between each retry
+   * @returns object {ndiPublicKeys, baseClient}
+   * @async
+   * @throws error if retrievePublicKeysFromNdi or retrieveBaseClientFromNdi fails
+   */
+  async createRefreshPromise(): Promise<Refresh> {
+    const [ndiPublicKeys, baseClient] = await retryPromiseForever(
+      () =>
+        Promise.all([
+          this.retrievePublicKeysFromNdi(),
+          this.retrieveBaseClientFromNdi(),
+        ]),
+      `Promise.all([this.retrievePublicKeysFromNdi(), this.retrieveBaseClientFromNdi()])`,
+    )
+
+    this._cache.set(NDI_PUBLIC_KEY_NAME, ndiPublicKeys) // No TTL - key will be kept forever until refresh is successful on expiry of EXPIRY_NAME key
+    this._cache.set(BASE_CLIENT_NAME, baseClient) // No TTL - key will be kept forever until refresh is successful on expiry of EXPIRY_NAME key
+    this._cache.set(EXPIRY_NAME, 'expiry', 3600) // set expiry key with TTL of 1 hour, to trigger refresh ahead (note that expiry check is done every 60s)
+    return { ndiPublicKeys, baseClient }
+  }
+
+  /**
+   * Returns stored refresh promise if it exists and is pending, or else
+   * calls createRefreshPromise(), saves the refresh promise and returns it
+   * @returns object {ndiPublicKeys, baseClient}
+   * @async
+   */
+  async refresh(): Promise<Refresh> {
+    // If promise does not exist, create and store the promise
+    if (!this._refreshPromise) {
+      this._refreshPromise = this.createRefreshPromise().finally(() => {
+        this._refreshPromise = undefined // Clean up once promise is fulfilled
+      })
+    }
+
+    // Return the refresh promise
+    return this._refreshPromise
+  }
+
+  /**
+   * Method to make network call to retrieve public JWKS from NDI
+   * Max of 3 back-to-back attemps with timeout of 3s per attempt as per NDI specs
+   * @async
+   * @returns NDI's public keys
+   * @throws JwkError if keys are not the correct shape
+   */
+  async retrievePublicKeysFromNdi(): Promise<CryptoKeys> {
+    const getJwksWithRetries = retryPromiseThreeAttempts(
+      () => axios.get<PublicJwks>(this.#spOidcNdiJwksEndpoint),
```

Comment:
Since the axios call is the actual interaction to the NDI service, the timeout specs should be specified on it (rather than on the `retryPromiseThreeAttempts()` wrapper). It will allow Axios to do the necessary cleanup.

Same comment for `retrieveBaseClientFromNdi()` below.

Reply:
added the 3s timeout for the axios call in for `retrievePublicKeysFromNdi`. However, in `retrieveBaseClientFromNdi`, `Issuer.discover()` uses  `fetch` in its underlying implementation, which doesn't provide a timeout. So I've kept the 3s timeout in `retryPromiseThreeAttempts()`